crates/subspace-service/src/dsn.rs

```diff
@@ -112,48 +133,132 @@ pub(crate) async fn start_dsn_archiver<Spawner>(
                 continue;
             }
         }
-        let keys_iter = (first_piece_index..)
-            .take(archived_segment.pieces.count())
-            .map(|idx| (idx, PieceIndexHash::from_index(idx)))
-            .map(|(idx, hash)| (idx, hash.to_multihash()));
 
         spawner.spawn(
             "segment-publishing",
             Some("subspace-networking"),
             Box::pin({
-                let node = node.clone();
-
-                async move {
-                    for ((_idx, key), piece) in keys_iter.zip(archived_segment.pieces.as_pieces()) {
-                        //TODO: restore announcing after https://github.com/libp2p/rust-libp2p/issues/3048
-                        // trace!(?key, ?idx, "Announcing key...");
-                        //
-                        // let announcing_result = node.start_announcing(key).await;
-                        //
-                        // trace!(?key, "Announcing result: {:?}", announcing_result);
-
-                        match node.put_value(key, piece.to_vec()).await {
-                            Ok(mut stream) => {
-                                if stream.next().await.is_some() {
-                                    trace!(?key, "Put value succeeded");
-                                } else {
-                                    trace!(?key, "Put value failed");
-                                }
-                            }
-                            Err(error) => {
-                                trace!(?key, "Put value failed: {}", error);
-                            }
-                        }
-
-                        //TODO: ensure republication of failed announcements
-                    }
-
-                    info!(%segment_index, "Segment processed.");
-                }
+                publish_pieces(
+                    node.clone(),
+                    first_piece_index,
+                    segment_index,
+                    archived_segment,
+                    piece_publisher_batch_size,
+                )
                 .in_current_span()
             }),
         );
 
         last_published_segment_index = Some(segment_index);
     }
 }
+
+// Publishes pieces-by-sector to DSN in bulk. Supports cancellation.
+pub(crate) async fn publish_pieces(
+    node: Node,
+    first_piece_index: PieceIndex,
+    segment_index: u64,
+    archived_segment: Arc<ArchivedSegment>,
+    piece_publisher_batch_size: usize,
+) {
+    let semaphore = Arc::new(Semaphore::new(piece_publisher_batch_size));
```

Comment:
This is correct for the farmer. The node doesn't use other batchers. Here is the issue for the farmer: https://github.com/subspace/subspace/issues/996

Reply:
What I meant there is not necessary a single semaphore for everything, but rather single semaphore for a specific task. Make it allow concurrency `10` for each getting and publishing, so that SDK that will use a single networking instance for everything will not fail. Should be quick and easy to do, I don't think we need to wait for another PR to fix such a small thing.